src/jdk.jfr/share/classes/jdk/jfr/internal/settings/Throttler.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.jfr.internal.settings;
+
+import java.util.Random;
+import java.util.concurrent.locks.ReentrantLock;
+import jdk.jfr.internal.PlatformEventType;
+public final class Throttler {
+    private static final ThrottlerParameters DISABLED_PARAMETERS = new ThrottlerParameters(0, 0, 0);
+    private static final long MILLIUNITS = 1000;
+    private static final long MINUTE = 60 * MILLIUNITS;
+    private static final long TEN_PER_1000_MS_IN_MINUTES = 600;
+    private static final long HOUR = 60 * MINUTE;
+    private static final long TEN_PER_1000_MS_IN_HOURS = 36000;
+    private static final long TEN_PER_1000_MS_IN_DAYS = 864000;
+    private static final long EVENT_THROTTLER_OFF = -2;
+
+    private final ReentrantLock lock = new ReentrantLock();
+    private final Random randomGenerator = new Random();
+    private final ThrottlerWindow window0 = new ThrottlerWindow();
+    private final ThrottlerWindow window1 = new ThrottlerWindow();
+
+    private volatile ThrottlerWindow activeWindow = window0;
+
+    // Guarded by lock
+    private double averagePopulationSize;
+    private double ewmaPopulationSize;
+    private long accumulatedDebtCarryLimit;
+    private long accumulatedDebtCarryCount;
+    private ThrottlerParameters lastParameters = new ThrottlerParameters(0, 0, 0);
+    private long sampleSize;
+    private long periodMillis;
+    private volatile boolean disabled;
+    private boolean update = true;
+
+    public Throttler(PlatformEventType t) {
+    }
+    // Not synchronized in fast path, but uses volatile reads.
+    public boolean sample(long ticks) {
+        if (disabled) {
```

Comment:
This volatile load is somewhat disappointing. Do you think it is needed? What happens if it read without happens-before? It just creates an event that will most likely get discarded by the recorder engine on reset (if its set on setting update). If it's set to disabled, then the recorder engine has most likely stopped already, so the event will be discarded. Event settings are set with no visibility guarantees as to exact when they apply, so it should not really matter when it goes to disabled.